dot/parachain/backing/per_relay_parent_state.go

```diff
@@ -14,19 +15,150 @@ import (
 	statementedistributionmessages "github.com/ChainSafe/gossamer/dot/parachain/statement-distribution/messages"
 	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
-	"github.com/ChainSafe/gossamer/dot/parachain/util"
+	parachainutil "github.com/ChainSafe/gossamer/dot/parachain/util"
 	"github.com/ChainSafe/gossamer/lib/common"
 )
 
 var errNilPersistedValidationData = errors.New("persisted validation data is nil")
 
+// constructPerRelayParentState constructs and returns the perRelayParentState for a given relay parent hash,
+// initialising various parameters and caches required for candidate backing.
+func (cb *CandidateBacking) constructPerRelayParentState(relayParent common.Hash) (*perRelayParentState, error) {
+	rt, err := cb.BlockState.GetRuntime(relayParent)
+	if err != nil {
+		return nil, fmt.Errorf("getting runtime for relay parent %s: %w", relayParent, err)
+	}
+
+	sessionIndex, err := rt.ParachainHostSessionIndexForChild()
+	if err != nil {
+		return nil, fmt.Errorf("getting session index: %w", err)
+	}
+
+	validators, err := cb.perSessionCache.getValidators(sessionIndex, rt)
+	if err != nil {
+		return nil, fmt.Errorf("getting validators: %w", err)
+	}
+
+	features, err := cb.perSessionCache.getNodeFeatures(sessionIndex, rt)
+	if err != nil {
+		return nil, fmt.Errorf("getting node features: %w", err)
+	}
+
+	if features == nil {
+		return nil, fmt.Errorf("invalid node features for relay parent %s", relayParent)
+	}
+
+	injectCoreIndex, err := features.Get(uint(parachaintypes.ElasticScalingMVP))
+	if err != nil {
+		return nil, fmt.Errorf("getting inject core index: %w", err)
+	}
+
+	executorParams, err := cb.perSessionCache.getExecutorParams(sessionIndex, rt)
+	if err != nil {
+		return nil, fmt.Errorf("getting executor params: %w", err)
+	}
+
+	validatorGroups, err := rt.ParachainHostValidatorGroups()
+	if err != nil {
+		return nil, fmt.Errorf("getting validator groups: %w", err)
+	}
+
+	minBackingVotes, err := cb.perSessionCache.getMinimumBackingVotes(sessionIndex, rt)
+	if err != nil {
+		return nil, fmt.Errorf("getting minimum backing votes: %w", err)
+	}
+
+	claimQueue, err := rt.ParachainHostClaimQueue()
+	if err != nil {
+		return nil, fmt.Errorf("getting claim queue: %w", err)
+	}
+
+	disabledValidators, err := rt.ParachainHostDisabledValidators()
+	if err != nil {
+		return nil, fmt.Errorf("getting disabled validators: %w", err)
+	}
+
+	signingContext := parachaintypes.SigningContext{
+		SessionIndex: sessionIndex,
+		ParentHash:   relayParent,
+	}
+
+	var localValidator *validator
+
+	validatorID, validatorIndex := parachainutil.SigningKeyAndIndex(validators, cb.Keystore)
+	if validatorID != nil {
+		//  local node is a validator
+		localValidator = &validator{
+			signingContext: signingContext,
+			key:            *validatorID,
+			index:          validatorIndex,
+			disabled:       slices.Contains(disabledValidators, validatorIndex),
+		}
+	}
+
+	numOfCores := uint32(len(validatorGroups.Validators))
```

Comment:
would this ever overflow?

Reply:
I don't think. because core index of of type `uint32`. So it's impossible to have a number of core more than `MaxUint32`.
```
type CoreIndex struct {
	Index uint32 `scale:"1"`
}
```